cmd/snap/cmd_run.go

```diff
@@ -482,26 +483,70 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+func getInfoAndApp(snapName, appName string, rev snap.Revision) (*snap.Info, *snap.AppInfo, error) {
+	info, err := getSnapInfo(snapName, rev)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	app := info.Apps[appName]
+	if app == nil {
+		return nil, nil, fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	}
+
+	return info, app, nil
+}
+
 func (x *cmdRun) snapRunApp(snapApp string, args []string) error {
 	if x.DebugLog {
 		os.Setenv("SNAPD_DEBUG", "1")
 		logger.Debugf("enabled debug logging of early snap startup")
 	}
 	snapName, appName := snap.SplitSnapApp(snapApp)
-	info, err := getSnapInfo(snapName, snap.R(0))
+
+	snapRev := snap.R(0)
+	hint, inhibitInfo, err := isLocked(snapName)
 	if err != nil {
 		return err
 	}
+	inhibited := hint != runinhibit.HintNotInhibited
+	if inhibited {
+		snapRev = inhibitInfo.Previous
+	}
 
-	app := info.Apps[appName]
-	if app == nil {
-		return fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	info, app, err := getInfoAndApp(snapName, appName, snapRev)
+	if err != nil {
+		return err
 	}
 
 	if !app.IsService() {
 		if err := maybeWaitWhileInhibited(snapName); err != nil {
 			return err
 		}
+
+		if inhibited {
+			// We were inhibited earlier, let's get updated "current" snap info
+			err = runinhibit.WithReadLock(snapName, func() error {
```

Comment:
Another nice thing: in this case, you'd only have to have this passed in as the `action`, in place of what you currently have when you use `WithReadLock`:

```go
func() error {
	info, app, err = getInfoAndApp(snapName, appName, snap.R(0))
	if err != nil {
		return err
	}
	return nil
}
```

This would also involve fewer concurrent calls to `IsLocked()`.

Reply:
Thanks for the review! This approach is definitely better from a UX perspective but I am worried it might be adding more complexity for a rare non-destructive edge-case.

@pedronis would love to know you thoughts and suggestions on this as well.

Reply:
although it's more complex I fear we might need this anyway to be correct, otherwise in the other approach we need to loop back into waitng. but notice that this would need actually to use NewFileLock... to be correct etc

Reply:
not saying that we should do this but ideally we would then keep the lock until we are executing or erroring or at least until we are added to the transient cgroup (see code in runSnapConfine for that)

Reply:
Things get a bit messy, but you could also do something like move the parts of `waitWhileInhibited` which call the status flow into the for proposed for loop as well, so if it sees that the inhibited status is refresh inhibited, it will do the flow, and if the snap becomes refresh inhibited again between when it is uninhibited and when the read lock can be acquired for performing the `action`, it could do the flow again.